### ISSUE TITLE
How to generate just a depency graph using xelatex.

### ISSUE THREAD

**Issue by ouboub** [PatrickMassot/leanblueprint#51]
Hi I am on Ubuntu 24.04 (either with tcsh or bash shell) I followed the instructions found in <https://leanprover-community.github.io/install/debian.html> and <https://leanprover-community.github.io/install/project.html>

And installed leanprover succesfully I am mainly interested, for the moment, in using blueprint, for my manuscripts, because I want go generate a dependency graph for the proof of theorems, using lemma definition etc. I wish I could use lean for my research, (non linear hyperbolic PDEs), but it seems that Lean, still is not suited for such a task.

So what I have in mind, was for example described in [Terence Tao: diagram of papers](https://terrytao.wordpress.com/2023/02/18/would-it-be-possible-to-create-a-tool-to-automatically-diagram-papers/)

Tao mentioned quiver which is nice, but you have to construct most of the graph yourself.

Then there is [rdfref](https://github.com/michal-h21/rdfref) But it is a bit buggy, and not that comfortable to use.

So I thought to give blueprint a try, but I run into problem, that I want to report.

Here is what I did

# Installation of blueprint

I also followed the instructions [How to install blueprint](https://github.com/PatrickMassot/leanblueprint?tab=readme-ov-file)

So in my my<sub>proyect</sub> folder which is

```
~/ALLES/HGs/User/Pub/Preprints/Lean/mathematics_in_lean/my_project/my_project 
```

After running `leanblueprint new` I have the folder blueprint with a couple of files (I don't understand why `\input` is used instead of `\subfile` which allows to run the content.tex file directly) in any case I just wanted to create a dependency graph

So I have

``` latex
\documentclass[a4paper]{report}
\usepackage{tikz-cd}
\usepackage{geometry}

\usepackage{expl3}
\usepackage[thms=dfn+lem+prop+thm+cor]{blueprint}
\usepackage{amssymb, amsthm, mathtools}
\usepackage[unicode,colorlinks=true,linkcolor=blue,urlcolor=magenta, citecolor=blue]{hyperref}

\usepackage[warnings-off={mathtools-colon,mathtools-overbracket}]{unicode-math}

\input{macros/common}
\input{macros/print}

\title{Energy estimates}
\author{Uwe Brauer}

\begin{document}
\maketitle
\input{content}
\end{document}
```

And the content file reads

``` latex
\begin{theorem}[Smale 1958]
  \label{thm:sphere_eversion}
  \lean{sphere_eversion}
  \leanok
  \uses{def:immersion}
  There is a homotopy of immersions of $𝕊^2$ into $ℝ^3$ from the inclusion map to
  the antipodal map $a : q ↦ -q$.
\end{theorem}


\begin{theorem}[Decaying energy estimate]
  \label{thm:open-ample}
  \lean{sphere_eversion}
  \leanok
  \uses{def:immersion}
  The following estimate holds
\end{theorem}

\begin{proof}
  \leanok
  \uses{thm:open-ample, lem:open_ample_immersion}
  This obviously follows from what we did so far.
\end{proof}
```

But no graph is generated, what do I miss, or where do I find relevant documentation? 

thanks

Uwe Brauer

**Comment by utensil** [comment]
Option `dep_graph` is missing from your `\usepackage[dep_graph]{blueprint}` line.

It will be there if you use the leanblueprint CLI tool to generate your project instead of manually setting it up.

By the way, leanblueprint works perfectly fine without lean. You can use only `\label`, `\uses`, and you'll already have your dependency graph.

**Comment by ouboub** [comment]
> Option `dep_graph` is missing from your `\usepackage[dep_graph]{blueprint}` line.
> 
> It should be there if you use the leanblueprint CLI tool to generate your project.
> 
> By the way, leanblueprint works perfectly fine without lean. You can use only `\label`, `\uses`, and you'll already have your dependency graph. `\leanok` can be used to mark the state, and it also doesn't involve lean except having lean in its name. If you are not using lean, you don't need to use `\lean`.

Well it does not work, sorry to say (TL 2004, xelatex)
content file reads

```

\begin{theorem}[Smale 1958]
  \label{thm:sphere_eversion}
  \lean{sphere_eversion}
  \leanok
  \uses{def:immersion}
  There is a homotopy of immersions of $𝕊^2$ into $ℝ^3$ from the inclusion map to
  the antipodal map $a : q ↦ -q$.
\end{theorem}


\begin{theorem}[Decaying energy estimate]
  \label{thm:open-ample}
  \lean{thm:open-ample}
  \leanok
  \uses{def:immersion}
  The following estimate holds
\end{theorem}

\begin{proof}
  \leanok
  \uses{thm:open-ample}
  This obviously follows from what we did so far.
\end{proof}
```
here is the pdf file 
[print.pdf](https://github.com/user-attachments/files/17485962/print.pdf)

**Comment by utensil** [comment]
Do you have the line like

```
\usepackage[showmore,dep_graph]{blueprint}
```

And I guess there is a misunderstanding here, dependency graph is only generated for the web version, see e.g.  https://imperialcollegelondon.github.io/FLT/blueprint/dep_graph_chapter_2.html .

However, if you wish to include the same graph in your LaTeX, you can copy the dot code from the source code of the html page (right click in the browser and choose view page, then scroll down to almost the bottom), something like:

```
.renderDot(`strict digraph "" {	graph ...lots of stuff .... } `)
```
 
Then you can copy everything between the backtick, and use [dot2tex](https://dot2tex.readthedocs.io/en/latest/) or similar.

It will look like this:

```
\begin{figure}[H]
    \centering
    {
    \begin{dot2tex}[dot,graphstyle={scale=0.8,transform shape}]
        strict digraph "" {	graph ...lots of stuff .... }
    \end{dot2tex}
    }
    \caption{Dependency graph}
    \label{fig:dep_graph}
\end{figure}
```

**Comment by ouboub** [comment]
>>> "U" == Utensil  ***@***.***> writes:

> Do you have the line like
> ```
> \usepackage[showmore,dep_graph]{blueprint}
> ```

No, in your last reply you only mentioned 
```
\usepackage[dep_graph]{blueprint}
``

> And I guess there is a misunderstanding here, dependency graph is only generated for the web version, see e.g.  https://imperialcollegelondon.github.io/FLT/blueprint/dep_graph_chapter_2.html .

Aha, my hope was that it could be used a bit like rdfref, I mentioned in my first message, but appearantly it can't

So if it is not too much to ask, how would I generate the web version (I presume it will only appear on my github repository but not say on a web page running apache on my Laptop)?

> However, if you wish to include the same graph in your LaTeX, you can copy the dot code from the source code of the html page (right click in the browser and choose view page, then scroll down to almost the bottom), something like:

> ```
> .renderDot(`strict digraph "" {	graph ...lots of stuff .... } `)
> ```
>  
> Then you can copy everything between the backtick, and use [dot2tex](https://dot2tex.readthedocs.io/en/latest/) or similar.

Ok dot2tex is what rdfref uses as well.


> It will look like this:

> ```
> \begin{figure}[H]
>     \centering
>     {
>     \begin{dot2tex}[dot,graphstyle={scale=0.8,transform shape}]
>         strict digraph "" {	graph ...lots of stuff .... }
>     \end{dot2tex}
>     }
>     \caption{Dependency graph}
>     \label{fig:dep_graph}
> \end{figure}
> ```

-- 
I strongly condemn Hamas heinous despicable pogroms/atrocities on Israel
I strongly condemn Putin's war of aggression against Ukraine.
I support to deliver weapons to Ukraine's military. 
I support the EU and NATO membership of Ukraine.

**Comment by ouboub** [comment]
>>> "U" == Utensil  ***@***.***> writes:

> Do you have the line like
> ```
> \usepackage[showmore,dep_graph]{blueprint}
> ```

No, in your last reply you only mentioned 
```
\usepackage[dep_graph]{blueprint}
``

> And I guess there is a misunderstanding here, dependency graph is only generated for the web version, see e.g.  https://imperialcollegelondon.github.io/FLT/blueprint/dep_graph_chapter_2.html .

Aha, my hope was that it could be used a bit like rdfref, I mentioned in my first message, but appearantly it can't

So if it is not too much to ask, how would I generate the web version (I presume it will only appear on my github repository but not say on a web page running apache on my Laptop)?

> However, if you wish to include the same graph in your LaTeX, you can copy the dot code from the source code of the html page (right click in the browser and choose view page, then scroll down to almost the bottom), something like:

> ```
> .renderDot(`strict digraph "" {	graph ...lots of stuff .... } `)
> ```
>  
> Then you can copy everything between the backtick, and use [dot2tex](https://dot2tex.readthedocs.io/en/latest/) or similar.

Ok dot2tex is what rdfref uses as well.


> It will look like this:

> ```
> \begin{figure}[H]
>     \centering
>     {
>     \begin{dot2tex}[dot,graphstyle={scale=0.8,transform shape}]
>         strict digraph "" {	graph ...lots of stuff .... }
>     \end{dot2tex}
>     }
>     \caption{Dependency graph}
>     \label{fig:dep_graph}
> \end{figure}
> ```

**Comment by utensil** [comment]
`showmore` is not required, I just didn't delete all my options.

The webpage is generated and pushed to your Github Pages by Github Actions workflows, as documented in README.

To see them on your own laptop, you don't need apache, just `leanblueprint serve` which is also mentioned in README.

**Comment by ouboub** [comment]
`leanblueprint serve` 
returned an error of a missing web directory. 
So I thought, I start again from scratch and I encountered some errors I can't remember to 
have seen before:

- I need lean installed, because running `leanblueprint new` leads to an error:

```
Error: Could not find a Lean project. Please run this command from inside your project folder.
```

- I then created a new empty public git repository, and enabled git actions in the pages section (but did not specify any action)

![githubactions](https://github.com/user-attachments/assets/272a6cfd-451d-444d-a522-3b73428f0f88)
- then i run  `leanblueprint new`  within my leanproject folder 
when asked about the url of the github repository I insert my new empty github repository , the same for the website project and for the docs...
I use the minimal Jekyell them, 
Now I receive 
```
 Git commit created. Don't forget to push when you are ready.

You are all set 🎉
```
but 
`git remote -v`
returns
```
origin  https://github.com/leanprover-community/mathematics_in_lean.git (fetch)
origin  https://github.com/leanprover-community/mathematics_in_lean.git (push)
```
Ah, what? I don't want and should not push to his repository, 
what do I miss  here?

**Comment by ouboub** [comment]
Indeed my first goal is to generate a graph similar to [chapter 2 Fermat](https://imperialcollegelondon.github.io/FLT/blueprint/dep_graph_chapter_2.html) **locally** using

-   leanbluprint web
-   leanbluprint server
-   and display it via my browser

So I tried out the **exact example** of [leanbluprint github](https://github.com/PatrickMassot/leanblueprint) namely

``` latex
\begin{theorem}[Smale 1958]
  \label{thm:sphere_eversion}
  \lean{sphere_eversion}
  \leanok
  \uses{def:immersion}
  There is a homotopy of immersions of $𝕊^2$ into $ℝ^3$ from the inclusion map to
  the antipodal map $a : q ↦ -q$.
\end{theorem}
  
\begin{proof}
  \leanok
  \uses{thm:open_ample, lem:open_ample_immersion}
  This obviously follows from what we did so far.
\end{proof}
```

and copy this into content.tex then running

-   leanbluprint web
-   leanbluprint server
-   and display it via my browser

Leads to the following display, which I attach,

![leanblueprint](https://github.com/user-attachments/assets/4fa93a60-48b8-4be1-aefe-99f08c5a7add)


 but which is not similar to the graph found in [chapter 2 Fermat](https://imperialcollegelondon.github.io/FLT/blueprint/dep_graph_chapter_2.html)

What do I miss?

Uwe Brauer

**Comment by utensil** [comment]
Your screenshot isn't complete, on the left panel there should be a chapter with name like dependency graph, and that's where the dep graph is.

**Comment by ouboub** [comment]
> Your screenshot isn't complete, on the left panel there should be a chapter with name like dependency graph, and that's where the dep graph is.

attached you find the complete screenshot, there is **no** left panel (using the latest chrome browser) (chrome displayed the Fermat example correctly!)
![leanblueprintcomplete](https://github.com/user-attachments/assets/3e2f8655-4c8c-4c7a-9e5d-76528bd87154)

**Comment by utensil** [comment]
It's hard to tell which step went wrong without looking at your code. By only descriptions of what you have been piecing things together, I can only make guesses, that's a long shot.

Instead, I would suggest to clone one of the example repos listed in the README, make them work locally, then reduce to what you need.

**Comment by ouboub** [comment]
ok, that  makes sense, shall I clone it into
```
~/ALLES/HGs/User/Pub/Preprints/Lean/mathematics_in_lean/my_project 
```
?

**Comment by utensil** [comment]
Anywhere that doesn't affect your exsiting files should be fine.

**Comment by ouboub** [comment]
ok, I did, that was helpful thanks

**Comment by ElifUskuplu** [comment]
@ouboub Hi, can you clarify how did you get the dependency graph locally? I would like to create one for my latex document.

**Comment by ouboub** [comment]
>>> "E" == ElifUskuplu  ***@***.***> writes:

> @ouboub Hi, can you clarify how did you get the dependency graph locally? I would like to create one for my latex document.

Well, honestly I gave up to use leanblueprint, because it consumed 10 GB of my disk and I found a more flexible solution namely 
https://github.com/michal-h21/rdfref?tab=readme-ov-file

Have a look in the issue section you find various examples discussed.

Per default a png is generated but you can also use `dot2tex` to generate directly a graph in LaTeX, using lualatex and shell-escape

The problem is if the graph is large (that is why leanblueprint uses a webserver. For LaTeX I use 

`\documentclass[preview,border={5cm 4cm 45cm 4cm},multi=true,multi={minipage}]{standalone}`

I also have to emphasize that I use it mostly for dependency graphs of
proofs not so much of whole documents.

You can pin me if you want to know more details.



> -- 
I strongly condemn Hamas heinous despicable pogroms/atrocities on Israel
I strongly condemn Putin's war of aggression against Ukraine.
I support to deliver weapons to Ukraine's military. 
I support the EU and NATO membership of Ukraine.

**Comment by ElifUskuplu** [comment]
@ouboub I deleted my previous comment since I solved my problem. I should have said that. There was a bug in [plastexdepgraph](https://github.com/PatrickMassot/plastexdepgraph/blob/master/README.md) package, and @PatrickMassot solved it (thanks again). As you said, thanks to dot2tex, I also created my independent depgraph latex file, so it works fine. So without leanblueprint, we can work with plastexdepgraph.

That said, I find it inappropriate to include political statements in a technical discussion, especially when they are unsolicited. Since you've brought it up, I want to make it clear that I stand against all forms of war crimes, genocide, violence, and injustice—regardless of who commits them. I believe discussions in professional and open-source communities should remain focused on their intended topics.
